### PR TITLE
added classifier and retrieval model

### PR DESCRIPTION
**Pull request by vshan**
Hey all,
I've added a retrieval model and a classifier model. The retrieval model answers the factual questions and our gotbot model the conversation - engaging questions. 

The natural question is to determine if user's query should go to the retrieval model or the gotbot model -- this is done using the classifier model. The classifier model takes the input and tells where the query should go to. 

As of now, this is the architecture of the chatbot:

1. Take input from user
2. Classify user query as fact-seeking or conversational
3. If fact-seeking, fetch answer from retrieval model
4. Else if it is conversational, fetch answer from gotbot model
5. Give output to user

## Conversational Model
This is our gotbot model that we have been working on, the sequence-to-sequence LSTM model. Improvements to this specifically, refer to issue #19 

The conversational model gives a personality and human-like replies which compliments nicely with the retrieval model for factual questions.

### Why retrieval model?
The goal of the retrieval model is to answer any question from a given FAQ set. For example, if given a set of say 100 FAQs, then _any_ input or query passed to the model _must_ match to any one of the FAQ questions. 

This is useful because in a closed-domain like 'Game of Thrones', we can have a finite set of knowledge-seeking questions and answers. examples: 'what is winterfell?', 'who is robb stark?' etc. And similar questions like 'who is robb?' can match to some question in the knowledge base.

This is a retrieval model, retrieving from a fixed set of questions (knowledge-base). Our conversational model, on the other hand, is a generative model which generates new responses, in the style of the data fed to it.

### How does it work? 

The model itself is similar to our sequence-to-sequence model. It's a Sequence Autoencoder, basically instead of passing (query, response) pairs to our seq2seq model, we pass (query, query) pairs so that the intermediate summarization vector is the vector representation of the query itself. After training the neural network, we get all the vectors from the queries.

In essence, this is a form of 'sentence' embedding just like our word embeddings. 

During prediction, the user's query is passed and we find the embedding sentence with the highest cosine similarity with user's query. 

### How to get data for model?

**Question Generation**:  The file `make_got_qa.py` in `wiki_crawler` folder crawls wiki and then generates questions from linking subheadings with headings and answers as the text that follows. This is the file which generates QA pairs from wiki site.

## Classifier Model
Classifier model is simply an LSTM-RNN connected to a 2-output softmax layer. Trained on queries from conversation model as one label and queries from retrieval model as another label. Simple classification to determine if query is factual or conversational.

### CODE REVIEW

**Comment by kaushiksk** [comment]
@vshan  any resources to understand the model/code ?

**Comment by vshan** [comment]
@kaushiksk ah, didn't know that. Will update. 

As for understanding the model, it is essentially a sequence autoencoder (https://blog.keras.io/building-autoencoders-in-keras.html), these embed the queries onto a space -- so during predict time we produce an embedding of user's query and find the most similar FAQ question using cosine similarity. 

I will also add more comments and documentation.
